src/org/traccar/database/DeviceManager.java

```diff
@@ -420,4 +427,46 @@ public void resetTotalDistance(DeviceTotalDistance deviceTotalDistance) throws S
             throw new IllegalArgumentException();
         }
     }
+
+    public void sendCommand(Command command) throws RecoverablePduException, UnrecoverablePduException,
+            SmppTimeoutException, SmppChannelException, InterruptedException {
+        if (command.getSms()) {
+            Position lastPosition = getLastPosition(command.getDeviceId());
+            if (lastPosition != null) {
+                Context.getServerManager().getProtocol(lastPosition.getProtocol())
+                        .sendSmsCommand(devicesById.get(command.getDeviceId()).getPhone(), command);
+            } else if (command.getType().equals(Command.TYPE_CUSTOM)) {
+                Context.getSmppManager().sendMessageSync(devicesById.get(command.getDeviceId()).getPhone(),
+                        command.getString(Command.KEY_DATA), true);
+            } else {
+                throw new RuntimeException("Command " + command.getType() + " is not supported");
+            }
+        } else {
+            ActiveDevice activeDevice = Context.getConnectionManager().getActiveDevice(command.getDeviceId());
+            if (activeDevice != null) {
+                activeDevice.sendCommand(command);
+            } else {
+                throw new RuntimeException("Device is not online");
+            }
+        }
+    }
+
+    public Collection<CommandType> getCommandTypes(long deviceId, boolean sms) {
+        List<CommandType> result = new ArrayList<>();
+        Position lastPosition = Context.getDeviceManager().getLastPosition(deviceId);
+        if (lastPosition != null) {
+            Collection<String> commands;
+            if (sms) {
+                commands = Context.getServerManager().getProtocol(lastPosition.getProtocol()).getSupportedSmsCommands();
```

Comment:
I think the code will be more readable if you get protocol once and then use ternary operator to get corresponding commands list.